tensorflow/core/kernels/dynamic_stitch_op.cc

```diff
@@ -20,9 +20,120 @@ limitations under the License.
 #include "tensorflow/core/framework/tensor.h"
 #include "tensorflow/core/kernels/bounds_check.h"
 
+#ifdef GOOGLE_CUDA
+#include "tensorflow/core/kernels/cuda_device_array.h"
+#endif // GOOGLE_CUDA
+
 namespace tensorflow {
 
-template <class T>
+typedef Eigen::ThreadPoolDevice CPUDevice;
+#ifdef GOOGLE_CUDA
+typedef Eigen::GpuDevice GPUDevice;
+#endif // GOOGLE_CUDA
+
+
+// Dynamic stitch function for CPUDevice.
+template <typename T>
+void DynamicStitchCPU(OpKernelContext* ctx,
+                const int32 first_dim_size,
+                const OpInputList& indices_inputs, const OpInputList& data_inputs,
+                Tensor* merged) {
+  auto merged_flat = merged->flat_outer_dims<T>();
+  const int slice_size = merged_flat.dimension(1);
+  for (int input_num = 0; input_num < indices_inputs.size(); input_num++) {
+    const Tensor& indices = indices_inputs[input_num];
+    auto indices_vec = indices.flat<int32>();
+    const Tensor& data = data_inputs[input_num];
+    auto data_flat =
+            data.shaped<T, 2>({indices_vec.dimension(0), slice_size});
+
+    if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
+      T* merged_base = &merged_flat(0, 0);
+      const T* data_base = &data_flat(0, 0);
+      const size_t slice_bytes = slice_size * sizeof(T);
+      for (int i = 0; i < indices_vec.size(); i++) {
+        int32 index = internal::SubtleMustCopy(indices_vec(i));
+        OP_REQUIRES(
+                ctx, FastBoundsCheck(index, first_dim_size),
+                errors::InvalidArgument("indices[", i, "] is out of range"));
+        memcpy(merged_base + index * slice_size, data_base + i * slice_size,
+               slice_bytes);
+      }
+    } else {
+      Eigen::DSizes<Eigen::DenseIndex, 2> sizes(1, slice_size);
+      for (int i = 0; i < indices_vec.size(); i++) {
+        // Copy slice data[i] to merged[indices[i]]
+        Eigen::DSizes<Eigen::DenseIndex, 2> data_indices(i, 0);
+        int32 index = internal::SubtleMustCopy(indices_vec(i));
+        OP_REQUIRES(
+                ctx, FastBoundsCheck(index, first_dim_size),
+                errors::InvalidArgument("indices[", i, "] is out of range"));
+        Eigen::DSizes<Eigen::DenseIndex, 2> merged_indices(index, 0);
+        merged_flat.slice(merged_indices, sizes) =
+                data_flat.slice(data_indices, sizes);
+      }
+    }
+  }
+}
+
+#ifdef GOOGLE_CUDA
+
+template <typename T>
+void DynamicStitchGPUImpl(const Eigen::GpuDevice& gpu_device,
+                          const int32 slice_size, const int32 first_dim_size,
+                          const CudaDeviceArrayStruct<int>& input_indices,
+                          const CudaDeviceArrayStruct<const T*>& input_ptrs,
+                          T* output);
+
+
+// because the collision requirements, we have to deal with
+// collion first before send data to gpu kernel.
+// TODO(ekelsen): This code can be done for more speed.
```

Comment:
How about - "Instead of doing a serial scan on the CPU to pick the last of duplicated indices, it could instead be done of the GPU implicitly using atomics to make sure the last index is the final write."